### ISSUE TITLE
Sniffing ignores authorization options and x-opaque-id

### ISSUE THREAD

**Issue by KACAH** [elastic/elasticsearch-py#2005]
**Elasticsearch version** (`bin/elasticsearch --version`):
Version: 8.2.2, Build: default/docker/9876968ef3c745186b94fdabd4483e01499224ef/2022-05-25T15:47:06.259735307Z, JVM: 18.0.1.1

**`elasticsearch-py` version (`elasticsearch.__versionstr__`)**:
8.2.3

Sniffing callback uses Transport instance directly to perform `/_nodes/_all/http` requests.
https://github.com/elastic/elasticsearch-py/blob/0da2ba901514447089f8a290f0ddba4f18cf53f8/elasticsearch/_async/client/_base.py#L172

However, both Authorization and opaque-id headers are NOT passed to Transport class during initialization.
https://github.com/elastic/elasticsearch-py/blob/0da2ba901514447089f8a290f0ddba4f18cf53f8/elasticsearch/_async/client/__init__.py#L324

Instead they are saved to Elasticsearch client instance after Transport is already created
https://github.com/elastic/elasticsearch-py/blob/0da2ba901514447089f8a290f0ddba4f18cf53f8/elasticsearch/_async/client/__init__.py#L407

and used for API requests only.
https://github.com/elastic/elasticsearch-py/blob/0da2ba901514447089f8a290f0ddba4f18cf53f8/elasticsearch/_async/client/__init__.py#L407 

As the result sniffing fails with Exception `elastic_transport.SniffingError: No viable nodes were discovered on the initial sniff attempt` while the real problem is

```json
{
   "error":{
      "root_cause":[
         {
            "type":"security_exception",
            "reason":"missing authentication credentials for REST request [/_nodes/_all/http]",
            "header":{
               "WWW-Authenticate":[
                  "Basic realm=\"security\" charset=\"UTF-8\"",
                  "ApiKey"
               ]
            }
         }
      ],
      "type":"security_exception",
      "reason":"missing authentication credentials for REST request [/_nodes/_all/http]",
      "header":{
         "WWW-Authenticate":[
            "Basic realm=\"security\" charset=\"UTF-8\"",
            "ApiKey"
         ]
      }
   },
   "status":401
}
```

The problem is the same for both Sync and Async clients.

The fix might be to build opaque-id and authorization headers before transport intialization to and pass them to `client_node_configs` method (`headers` argument). However, I am not sure if it will work "as intended" with `.options()` feature and internal `_transport` constructor argument of `Elasticsearch` and `AsyncElasticsearch` classes.

**Comment by VanillaDevelop** [comment]
This appears to still be an issue on elasticsearch-py version 8.5.0 and elasticsearch version 8.5.3. The misleading error message cost us quite some time re-evaluating our elasticsearch configuration. Would be nice if someone could take a look at it.

**Comment by siyanew** [comment]
Is there any way to have `sniff_on_start=True` and send authorization to sniff request? Any update or fix?

**Comment by jhopkins219** [comment]
Howdy, I ran into this problem upgrading our clients from 7.x to 8.8 and wrote a workaround by overwriting the internal callback that does these sniff requests. Hope this issue can be addressed soon, but in the meantime you can still use the sniff options by initializing the connection like so (note this is for the sync client, and may need changes depending on your client version and needs):

https://gist.github.com/jhopkins219/2b5b38061f1b87515c7a5c29ca91a0a3

**Comment by acidvegas** [comment]
Bumping cause this 2022 issue is still persistent.....classic enterprise devs lol

**Comment by acidvegas** [comment]
> Howdy, I ran into this problem upgrading our clients from 7.x to 8.8 and wrote a workaround by overwriting the internal callback that does these sniff requests. Hope this issue can be addressed soon, but in the meantime you can still use the sniff options by initializing the connection like so (note this is for the sync client, and may need changes depending on your client version and needs):
> 
> https://gist.github.com/jhopkins219/2b5b38061f1b87515c7a5c29ca91a0a3

I can confirm @jhopkins219 patch fixes this issue for me! Props

**Comment by prsinghs** [comment]
is this fix released? facing the same on 8.12 elasticsearch-py

**Comment by acidvegas** [comment]
Crazy that this is an issue from 2022 and we have a proof-of-concept fix for it, yet it hasnt been pushed to the upstream at all lol.

**Comment by acidvegas** [comment]
> is this fix released? facing the same on 8.12 elasticsearch-py

nope. lazy enterprise devs still havent paid attention to this 2 year old issue.....

https://github.com/elastic/elasticsearch-py/blob/8.12/elasticsearch/_sync/client/_base.py#L166
^ missing the auth header entirely still. Use the patch @jhopkins219 shared, it works perfectly.

Literally this issue is a 1 line fix. Blows my mind this is from 2022 and still not pushed to the upstream.
